parlai/tasks/taskmaster2/agents.py

```diff
@@ -151,58 +152,42 @@ def custom_evaluation(
-            correct = 0
-            for slot_guess in slot_guesses:
-                if ' = ' not in slot_guess:
-                    continue
-                try:
-                    slot, guess = slot_guess.split(' = ')
-                except ValueError:
-                    continue
-                if teacher_action['slots'].get(slot) == guess:
-                    self.metrics.add('slot_p', AverageMetric(1))
-                    self.metrics.add(f'{domain}_slot_p', AverageMetric(1))
-                    correct += 1
-                else:
-                    self.metrics.add('slot_p', AverageMetric(0))
-                    self.metrics.add(f'{domain}_slot_p', AverageMetric(0))
-                    logging.debug(
-                        f"Bad slot guess '{slot_guess}' != {teacher_action['slots']}"
-                    )
-            if teacher_action['slots']:
-                self.metrics.add(
-                    'slot_r', AverageMetric(correct, len(teacher_action['slots']))
-                )
-                self.metrics.add(
-                    f'{domain}_slot_r',
-                    AverageMetric(correct, len(teacher_action['slots'])),
+            slot_guesses = dict(
+                map(
+                    lambda x: x.split(' = '),
+                    filter(lambda x: ' = ' in x, slot_guesses_raw),
                 )
-                self.metrics.add(
-                    'jga', AverageMetric(correct == len(teacher_action['slots']))
+            )
+
+            self.metrics.add_metrics(
+                SlotMetrics(
+                    teacher_slots=teacher_action.get('slots', {}),
+                    count_empty_teacher=teacher_action['slots'],
+                    predicted_slots=slot_guesses,
+                    domain=domain,
+                    valid_domains=DOMAINS,
                 )
+            )
 
         elif teacher_action['type'] == 'apiresp':
-            # keep track of statistics by domain
-            f1_metric = F1Metric.compute(model_response['text'], labels)
-            bleu_metric = BleuMetric.compute(model_response['text'], labels)
-            self.metrics.add(f'{domain}_lex_f1', f1_metric)
-            self.metrics.add(f'{domain}_lex_bleu', bleu_metric)
-
             delex_text = model_response['text']
             delex_label = labels[0]
             # compute delexicalized string metrics
             for slot, value in teacher_action['slots'].items():
                 delex_text = delex_text.replace(value, slot)
                 delex_label = delex_label.replace(value, slot)
-            f1_metric = F1Metric.compute(delex_text, (delex_label,))
-            self.metrics.add('delex_f1', f1_metric)
-            self.metrics.add(f'{domain}_delex_f1', f1_metric)
-            bleu_metric = BleuMetric.compute(delex_text, [delex_label])
-            self.metrics.add('delex_bleu', bleu_metric)
-            self.metrics.add(f'{domain}_delex_bleu', bleu_metric)
+            self.metrics.add_metrics(
```

Comment:
Okay I like this. I just think there's no implementation of add_metrics.

Reply:
It's in the "metrics_collections" branch ("slot_teachers" is rebased onto that, while this one is rebased on "slot_teachers").

This may be uh. Me trying to recreate diff stacks from Mecurial when that's not how things work in git land. Maybe the lesson I should be learning here is to not try to mirror diff stacks by using branches on top of branches... 🙃

Reply:
For context, this is the branch that includes the `add_metrics`: https://github.com/facebookresearch/ParlAI/pull/3145

(Given Github's UI... next time, perhaps I'll just do all of these changes in one pull request, since it's not super nice on stacked branches.)

Reply:
Ah yeah, stacked diffs aren't a thing in github. Sometimes I do the PRs on top of PRs but generally not.

There's a tool called ghstack that the pytorch people use to emulate fairly successfully.